week-06-intro-ruby-on-rails/day-02/module-01/readme.md

```diff
@@ -1 +1,445 @@
+
+#### Parameters can have default values
+
+```ruby
+def say(something = "Hello")
+  puts something
+end
+
+say
+# Hello
+# => nil
+
+say "Goodbye"
+# Goodbye
+# => nil
+```
+
+#### Recursion: methods can call themselves
+
+```ruby
+def recurse(depth)
+  if depth > 0
+    puts "#{depth} spiraling down..."
+    recurse(depth - 1)
+    puts "#{depth} spiraling up..."
+  else
+    puts "Bottom of the rabbit hole"
+  end
+end
+
+recurse 3
+
+# 3 spiraling down...
+# 2 spiraling down...
+# 1 spiraling down...
+# Bottom of the rabbit hole
+# 1 spiraling up...
+# 2 spiraling up...
+# 3 spiraling up...
+# => nil
+```
+
+#### Functions have locally scoped variables (biggest difference from JavaScript!)
+
+The following code won't change `foo`. Why?
+
+```ruby
+foo = 1
+
+def do_stuff
+  foo += 1
+  bar = 1
+  puts foo
+  puts bar
+end
+
+do_stuff
+# undefined method `+' for nil:NilClass
+
+puts foo
+# 1
+# => nil
+
+puts bar
+# undefined local variable or method `bar' for main:Object
+```
+
+The problem is that ruby is *entirely* locally scoped -- code within a function only has access to the function's parameters and any variables defined inside of the function.
+
+```ruby
+foo = 1
+
+def do_stuff
+  foo = 1
+  foo += 1
+  bar = 1
+  puts foo
+  puts bar
+end
+
+do_stuff
+# 2
+# 1
+# => nil
+
+puts foo
+# 1
+# => nil
+
+puts bar
+# undefined local variable or method `bar' for main:Object
+
+def do_other_stuff(x)
+  foo = x
+  foo += 1
+  bar = 1
+  puts foo
+  puts bar
+end
+
+puts do_other_stuff(foo)
+# 2
+# 1
+# => nil
+
+puts foo
+# 1
+# => nil
+```
+
+Block scoping vs functional scoping...
+
+###Example: Factorial
+
+A simple, common interview question asks candidates to compute the *factorial* of a number. (The factorial of a number just multiplies it by all lower positive numbers. For example, `6` factorial (written `6!`) is  `6*5*4*3*2*1`, which equals `720`.)
+
+
+```ruby
+def factorial(num)
+   if num > 1
+      num * factorial(num-1)
+   elsif num == 1 or num == 0
+      1
+   else
+      puts "can't do factorial of a negative number!"
+   end
+end
+
+factorial(6)
+# => 720
+
+factorial(-1)
+# can't do factorial of a negative number!
+# => nil
+```
+
+## How to run a ruby script inside a file
+1. `touch script.rb`
+2. `subl script.rb`
+3. `ruby script.rb`
+4. THAT'S IT!
+5. From irb you can run a ruby file by loading it `load './script.rb'`
+
+## How to use gems in irb
+
+Gems in Ruby are a lot like node modules. Say you want to use a gem like "Awesome Print" to make your printing of objects and strings colorized and indented?
+
+1. First install the gem on your computer: `gem install awesome_print`
+2. Now just require it at the top of your file or in irb and then use it according to its documentation.
+  ```ruby
+  require 'awesome_print'
+  a = ["value1", "value2", "value3"]
+  ap a
+  ```
+
+[Ruby Toolbox](https://www.ruby-toolbox.com/)
+
+##Basic Challenges
```

Comment:
NOOOOOOO. I HATE "Basic" Challenges. :grimacing:  Sorry. I know, not your fault.